command/oauth/cmd.go

```diff
@@ -702,6 +765,114 @@ func (o *oauth) DoManualAuthorization() (*token, error) {
 	return tok, nil
 }
 
+type identifyDeviceResponse struct {
+	DeviceCode      string `json:"device_code"`
+	UserCode        string `json:"user_code"`
+	VerificationURI string `json:"verification_uri"`
+	// NOTE Google returns `verification_url` which is incorrect
+	// according to the spec (https://datatracker.ietf.org/doc/html/rfc8628#section-3.2)
+	// but we'll try to accomodate for that here.
+	VerificationURL         string `json:"verification_url"`
+	VerificationURIComplete string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	Interval                int    `json:"interval"`
+}
+
+// DoDeviceAuthorization gets a token from the IDP using the OAuth 2.0
+// Device Authorization Grant. https://datatracker.ietf.org/doc/html/rfc8628
+func (o *oauth) DoDeviceAuthorization() (*token, error) {
+	// Identify the Device
+	data := url.Values{}
+	data.Set("client_id", o.clientID)
+	data.Set("client_secret", o.clientSecret)
+	data.Set("scope", o.scope)
+
+	resp, err := http.PostForm(o.deviceAuthzEndpoint, data)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	defer resp.Body.Close()
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode >= 400 {
+		var e struct {
+			Error string
+		}
+		if err := json.NewDecoder(bytes.NewReader(b)).Decode(&e); err != nil {
+			return nil, errors.Wrapf(err, "could not parse http body: %s", string(b))
+		}
+	}
+
+	var idr identifyDeviceResponse
+	if err := json.NewDecoder(bytes.NewReader(b)).Decode(&idr); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	switch {
+	case idr.VerificationURI != "":
+		break
+	case idr.VerificationURL != "":
+		// NOTE this is a hack for Google, because their API returns the attribute
+		// 'verification_url` rather than `verification_uri`.
+		idr.VerificationURI = idr.VerificationURL
+	default:
+		return nil, errors.Errorf("device code response from server missing 'verification_uri' parameter. http body response: %s", string(b))
+	}
+
+	fmt.Fprintln(os.Stderr, "Go to the following website:")
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprintln(os.Stderr, idr.VerificationURI)
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprintln(os.Stderr, "and enter the activation code below:")
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprintln(os.Stderr, idr.UserCode)
+	fmt.Fprintln(os.Stderr)
+
+	// Poll the Token endpoint until the user completes the flow.
+	data = url.Values{}
+	data.Set("client_id", o.clientID)
+	data.Set("client_secret", o.clientSecret)
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+	data.Set("device_code", idr.DeviceCode)
+
+	var tok token
+	for {
+		resp, err := http.PostForm(o.tokenEndpoint, data)
```

Comment:
We should probably sleep a few seconds before doing the first request. In those milliseconds between printing the message and doing the request, there's not enough time no matter how fast you are.